Sources/XCHammer/BazelQueryer.swift

```diff
@@ -65,36 +65,26 @@ extension String: BazelQueryLang {
 }
 
 enum BazelQueryer {
-    static func sourceFileDepsQueryCommand(targets: [BuildLabel], bazelPath: Path, workspaceRoot: Path) -> String {
-        let query = targets.flatMap{ target in
-            .kind(.str("source file"), .deps(.word(target.description)))
-        }.reduce(.set([]), String.union)
-        let command: String = ([
-            bazelPath.string,
-            "query",
-            "--output=label_kind",
-            "'\(query)'"
-        ]).joined(separator: " ")
-        return command
-    }
-    
     /// query bazel for any generated files the project depends on
     static func genFileQuery(targets: [BuildLabel], bazelPath: Path, workspaceRoot: Path) throws -> [BuildLabel] {
         let genfileProfiler = XCHammerProfiler("query_genfiles")
         defer {
             genfileProfiler.logEnd(true)
         }
-        let kindWhitelist: [String] = [
-            "entitlements",
-            "module_map",
-            "headermap"
-        ]
 
-        let query = targets.flatMap{ target in
-            kindWhitelist.map{ rule in
-                .kind(.str(rule), .deps(.word(target.description)))
-            }
-        }.reduce(.set([]), String.union)
+        // - include genrules
+        // - include rules of tag 'xchammer'. The user can include arbitrary
+        //   types of rules to be built for Xcode usage.
+        //   These files are assumed to be requsites of Xcode.
+        let queryParts: [String] = targets.reduce([]) {
+            accum, target -> [String] in
+            return accum + [
+                String.kind(.str("genrule"), .deps(.word(target.description))),
+                String.application(name: "attr", params: [.str("tags"),
+                        .str("xchammer"), .deps(.word(target.description))]),
+            ]
+        }
+        let query = queryParts.reduce(.set([]), String.union)
```

Comment:
I believe `genrule`s don't need the xchammer tag, it's other rules you want to opt in here.

Reply:
In hindsight, having both `genrule` and `xchammer` tag isn't useful. The user should manually opt in all rules that build ahead ahead of time ( both here and and during generation ).

I think we should have the option to defer code generation and downloads of required files until build time so that project generation is very fast for users. I submitted a [PR to XcodeGen](https://github.com/yonaskolb/XcodeGen/pull/410) to achieve this.

Reply:
@rahul-malik I'm going to submit a few more commits to add documentation, talks, and info about XCHammer